File: app/src/main/java/com/josycom/mayorjay/flowoverstack/ui/WebViewActivity.java
package com.josycom.mayorjay.flowoverstack.ui;

import androidx.annotation.ColorInt;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.webkit.WebChromeClient;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.ProgressBar;

import com.josycom.mayorjay.flowoverstack.R;
import com.josycom.mayorjay.flowoverstack.util.AppUtils;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Objects;

import static com.josycom.mayorjay.flowoverstack.util.StringConstants.WEBVIEW_EXTRA_OBJECT;

public class WebViewActivity extends AppCompatActivity {

    private String url;
    private WebView webView;
    private ActionBar actionBar;
    private ProgressBar progressBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_web_view);

        // get extra object
        url = getIntent().getStringExtra(WEBVIEW_EXTRA_OBJECT);

        initComponent();
        initToolbar();
        loadWebFromUrl();
    }

    private void initComponent() {
        webView = findViewById(R.id.webView);
        progressBar = findViewById(R.id.progressBar);
        progressBar.getProgressDrawable().setColorFilter(getResources().getColor(R.color.colorPrimaryLight), PorterDuff.Mode.SRC_IN);
        progressBar.setBackgroundColor(Color.parseColor("#1A000000"));
    }

    private void initToolbar() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setNavigationIcon(R.drawable.ic_clear);
        Objects.requireNonNull(toolbar.getNavigationIcon()).setColorFilter(getResources().getColor(R.color.colorPrimary), PorterDuff.Mode.SRC_ATOP);
        actionBar = getSupportActionBar();
        Objects.requireNonNull(actionBar).setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeButtonEnabled(true);
        actionBar.setTitle(null);
        changeOverflowMenuIconColor(toolbar, getResources().getColor(R.color.colorPrimary));
        setSystemBarColor(this, "#f2f2f2");
        setSystemBarLight(this);
    }

    @SuppressLint("SetJavaScriptEnabled")
    private void loadWebFromUrl() {
        webView.getSettings().setJavaScriptEnabled(true);
        webView.getSettings().setDefaultTextEncodingName("utf-8");
        webView.getSettings().setBuiltInZoomControls(true);
        webView.getSettings().setDisplayZoomControls(false);
        webView.setBackgroundColor(Color.TRANSPARENT);
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                super.onPageStarted(view, url, favicon);
                actionBar.setTitle(null);
                actionBar.setSubtitle(getHostName(url));
                progressBar.setVisibility(View.VISIBLE);
            }

            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                actionBar.setTitle(view.getTitle());
                progressBar.setVisibility(View.INVISIBLE);
            }
        });
        webView.loadUrl(url);
        webView.setWebChromeClient(new WebChromeClient() {
            public void onProgressChanged(WebView view, int progress) {
                progressBar.setProgress(progress + 10);
                if (progress >= 100) actionBar.setTitle(view.getTitle());
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
        } else if (item.getItemId() == R.id.action_refresh) {
            loadWebFromUrl();
        } else if (item.getItemId() == R.id.action_browser) {
            AppUtils.directLinkToBrowser(this, url);
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        getMenuInflater().inflate(R.menu.menu_webview, menu);
        changeMenuIconColor(menu, getResources().getColor(R.color.colorPrimary));
        return true;
    }

    @Override
    public void onBackPressed() {
        if (webView.canGoBack()) {
            webView.goBack();
        } else {
            super.onBackPressed();
        }
        overridePendingTransition(R.anim.fade_in_anim, R.anim.fade_out_anim);
    }

    @Override
    protected void onResume() {
        webView.onResume();
        super.onResume();
    }

    @Override
    protected void onPause() {
        webView.onPause();
        super.onPause();
    }

    public static void setSystemBarLight(Activity act) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            View view = act.findViewById(android.R.id.content);
            int flags = view.getSystemUiVisibility();
            flags |= View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
            view.setSystemUiVisibility(flags);
        }
    }

    public void setSystemBarColor(Activity act, String color) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            Window window = act.getWindow();
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
            window.setStatusBarColor(Color.parseColor(color));
        }
    }

    private void changeOverflowMenuIconColor(Toolbar toolbar, @ColorInt int color) {
        try {
            Drawable drawable = toolbar.getOverflowIcon();
            Objects.requireNonNull(drawable).mutate();
            drawable.setColorFilter(color, PorterDuff.Mode.SRC_ATOP);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void changeMenuIconColor(Menu menu, @ColorInt int color) {
        for (int i = 0; i < menu.size(); i++) {
            Drawable drawable = menu.getItem(i).getIcon();
            if (drawable == null) continue;
            drawable.mutate();
            drawable.setColorFilter(color, PorterDuff.Mode.SRC_ATOP);
        }
    }

    public String getHostName(String url) {
        try {
            URI uri = new URI(url);
            String new_url = uri.getHost();
            if (!new_url.startsWith("www.")) new_url = "www." + new_url;
            return new_url;
        } catch (URISyntaxException e) {
            e.printStackTrace();
            return url;
        }
    }

    public static void navigate(Activity activity, String url) {
        Intent i = navigateBase(activity, url);
        activity.startActivity(i);
    }

    public static Intent navigateBase(Context context, String url) {
        Intent intent = new Intent(context, WebViewActivity.class);
        intent.putExtra(WEBVIEW_EXTRA_OBJECT, url);
        return intent;
    }
}

File: app/src/main/java/com/josycom/mayorjay/flowoverstack/ui/activity/WebViewActivity.kt
package com.josycom.mayorjay.flowoverstack.ui.activity

import androidx.appcompat.app.AppCompatActivity
import android.webkit.WebView
import android.widget.ProgressBar
import android.os.Bundle
import com.josycom.mayorjay.flowoverstack.R
import com.josycom.mayorjay.flowoverstack.util.AppConstants
import android.graphics.PorterDuff
import android.annotation.SuppressLint
import android.webkit.WebViewClient
import android.graphics.Bitmap
import android.webkit.WebChromeClient
import android.app.Activity
import android.content.Context
import android.view.WindowManager
import androidx.annotation.ColorInt
import android.content.Intent
import android.graphics.Color
import android.os.Build
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.ActionBar
import androidx.appcompat.widget.Toolbar
import com.josycom.mayorjay.flowoverstack.util.AppUtils
import java.lang.Exception
import java.net.URI
import java.net.URISyntaxException

class WebViewActivity : AppCompatActivity() {

    private var url: String? = null
    private lateinit var webView: WebView
    private lateinit var actionBar: ActionBar
    private lateinit var progressBar: ProgressBar

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_web_view)

        // get extra object
        url = intent.getStringExtra(AppConstants.WEBVIEW_EXTRA_OBJECT)
        initComponent()
        initToolbar()
        loadWebFromUrl()
    }

    private fun initComponent() {
        webView = findViewById(R.id.webView)
        progressBar = findViewById(R.id.progressBar)
        progressBar.progressDrawable.setColorFilter(resources.getColor(R.color.colorPrimaryLight), PorterDuff.Mode.SRC_IN)
        progressBar.setBackgroundColor(Color.parseColor("#1A000000"))
    }

    private fun initToolbar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        toolbar.setNavigationIcon(R.drawable.ic_clear)
        toolbar.navigationIcon!!.setColorFilter(resources.getColor(R.color.colorPrimary), PorterDuff.Mode.SRC_ATOP)
        actionBar = supportActionBar!!
        actionBar.setDisplayHomeAsUpEnabled(true)
        actionBar.setHomeButtonEnabled(true)
        actionBar.title = null
        changeOverflowMenuIconColor(toolbar, resources.getColor(R.color.colorPrimary))
        setSystemBarColor(this)
        setSystemBarLight(this)
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun loadWebFromUrl() {
        webView.apply {
            settings.javaScriptEnabled = true
            settings.defaultTextEncodingName = "utf-8"
            settings.builtInZoomControls = true
            settings.displayZoomControls = false
            setBackgroundColor(Color.TRANSPARENT)
        }
        webView.webViewClient = object : WebViewClient() {
            override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
                super.onPageStarted(view, url, favicon)
                actionBar.title = null
                actionBar.subtitle = getHostName(url)
                progressBar.visibility = View.VISIBLE
            }

            override fun onPageFinished(view: WebView, url: String) {
                super.onPageFinished(view, url)
                actionBar.title = view.title
                progressBar.visibility = View.INVISIBLE
            }
        }
        webView.loadUrl(url!!)
        webView.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: WebView, progress: Int) {
                progressBar.progress = progress + 10
                if (progress >= 100) actionBar.title = view.title
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
            }
            R.id.action_refresh -> {
                loadWebFromUrl()
            }
            R.id.action_browser -> {
                AppUtils.directLinkToBrowser(this, url)
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_webview, menu)
        changeMenuIconColor(menu, resources.getColor(R.color.colorPrimary))
        return true
    }

    override fun onBackPressed() {
        if (webView.canGoBack()) {
            webView.goBack()
        } else {
            super.onBackPressed()
        }
    }

    override fun onResume() {
        webView.onResume()
        super.onResume()
    }

    override fun onPause() {
        webView.onPause()
        super.onPause()
    }

    private fun setSystemBarColor(act: Activity) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val window = act.window
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
            window.statusBarColor = Color.parseColor("#f2f2f2")
        }
    }

    private fun changeOverflowMenuIconColor(toolbar: Toolbar, @ColorInt color: Int) {
        try {
            val drawable = toolbar.overflowIcon
            drawable?.mutate()
            drawable!!.setColorFilter(color, PorterDuff.Mode.SRC_ATOP)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun getHostName(url: String): String {
        return try {
            val uri = URI(url)
            var newUrl = uri.host
            if (!newUrl.startsWith("www.")) newUrl = "www.$newUrl"
            newUrl
        } catch (e: URISyntaxException) {
            e.printStackTrace()
            url
        }
    }

    companion object {
        fun setSystemBarLight(act: Activity) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                val view = act.findViewById<View>(android.R.id.content)
                var flags = view.systemUiVisibility
                flags = flags or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
                view.systemUiVisibility = flags
            }
        }

        fun changeMenuIconColor(menu: Menu, @ColorInt color: Int) {
            for (i in 0 until menu.size()) {
                val drawable = menu.getItem(i).icon ?: continue
                drawable.mutate()
                drawable.setColorFilter(color, PorterDuff.Mode.SRC_ATOP)
            }
        }

        fun navigate(activity: Activity, url: String?) {
            val i = getIntent(activity, url)
            activity.startActivity(i)
        }

        private fun getIntent(context: Context?, url: String?): Intent {
            Intent(context, WebViewActivity::class.java).apply {
                putExtra(AppConstants.WEBVIEW_EXTRA_OBJECT, url)
                return this
            }
        }
    }
}